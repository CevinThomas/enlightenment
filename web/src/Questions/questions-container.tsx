import Axios from "axios";
import React, { useRef, useState } from "react";
import { Redirect } from "react-router-dom";
import Cookies from "universal-cookie";
import EnvVariables from "../envVariables";
import { QuestionsData } from "../interfaces/questionsData";
import QuestionsView from "./questions-view";

const QuestionsContainer = () => {
  const [addedQuestions, setAddedQuestions] = useState<Array<QuestionsData>>(
    []
  );
  const [editingQuestionState, setEditingQuestionState] = useState<boolean>(
    false
  );
  const [groupName, setGroupName] = useState<string>("");
  const [areaName, setAreaName] = useState<string>("");
  const [subjectName, setSubjectName] = useState<string>("");
  const [category, setCategory] = useState<string>("");

  const [questionsData, setQuestionsData] = useState({
    id: 0,
    name: "",
    category: "",
    groupName: "",
    areaName: "",
    subjectName: "",
    groupId: 0,
    answered: 0,
    answeredCorrectly: 0,
    options: [
      { choice: "", correct: false, explanation: "", id: 0 },
      { choice: "", correct: false, explanation: "", id: 1 },
      { choice: "", correct: false, explanation: "", id: 2 },
      { choice: "", correct: false, explanation: "", id: 3 },
    ],
    tags: "",
  });

  const counterForQuestions = useRef<number>(0);

  function deleteQuestion(id: number): void {
    const clonedSavedQuestions = [...addedQuestions];
    const filteredQuestions = clonedSavedQuestions.filter(
      (question) => question.id !== id
    );
    setAddedQuestions(filteredQuestions);
  }

  function saveQuestion(): void {
    const addedQuestionsClone = [...addedQuestions];
    //TODO Validate questions before adding
    addedQuestionsClone.push(questionsData);
    setAddedQuestions(addedQuestionsClone);
    counterForQuestions.current = counterForQuestions.current + 1;
    resetQuestionsData();
  }

  function updateEditingQuestion(id: number): void {
    const clonedAddedQuestions = [...addedQuestions];

    for (let i = 0; i < clonedAddedQuestions.length; i++) {
      if (clonedAddedQuestions[i].id === questionsData.id) {
        clonedAddedQuestions[i] = questionsData;
      }
    }

    setAddedQuestions(clonedAddedQuestions);

    resetQuestionsData();
    setEditingQuestionState(false);
  }

  function cancelUpdateEditingQuestion(id: number): void {
    setEditingQuestionState(false);
    resetQuestionsData();
  }

  function editQuestion(id: number): void {
    const clonedSavedQuestions = [...addedQuestions];
    const questionToEdit = clonedSavedQuestions.filter(
      (question) => question.id === id
    );
    setQuestionsData(questionToEdit[0]);
    setEditingQuestionState(true);
  }

  async function submitQuestions() {
    const clonedAddedQuestions = [...addedQuestions];

    clonedAddedQuestions.forEach((question) => {
      question.groupName = groupName;
      question.areaName = areaName;
      question.subjectName = subjectName;
      question.category = category;
    });

    const cookies = new Cookies();

    const response = await Axios.post(
      `${EnvVariables.API_ENDPOINTS.ADDQUESTIONS}`,
      clonedAddedQuestions,
      {
        headers: {
          Authorization: cookies.get("token"),
        },
      }
    );

    if (response.data.data.statusCode === 200) {
      return <Redirect to="/all-questions" />;
    }
  }

  function resetQuestionsData(): void {
    setQuestionsData({
      id: counterForQuestions.current,
      name: "",
      category: "",
      groupName: "",
      areaName: "",
      subjectName: "",
      groupId: 0,
      answered: 0,
      answeredCorrectly: 0,
      options: [
        { choice: "", correct: false, explanation: "", id: 0 },
        { choice: "", correct: false, explanation: "", id: 1 },
        { choice: "", correct: false, explanation: "", id: 2 },
        { choice: "", correct: false, explanation: "", id: 3 },
      ],
      tags: "",
    });
  }

  function onOptionExplanationHandlerChange(e: any): void {
    const clonedOptions = questionsData.options;
    clonedOptions[e.target.name].explanation = e.target.value;
    setQuestionsData({
      ...questionsData,
      options: clonedOptions,
    });
  }

  function areaNameChange(e: any): void {
    setAreaName(e.target.value);
  }

  function subjectNameChange(e: any): void {
    setSubjectName(e.target.value);
  }

  function categoryNameChange(e: any): void {
    setCategory(e.target.value);
  }

  function onNameChangeHandler(e: any): void {
    setQuestionsData({
      ...questionsData,
      name: e.target.value,
    });
  }

  function onTagChangeHandler(e: any): void {
    setQuestionsData({
      ...questionsData,
      tags: e.target.value,
    });
  }

  function onCorrectChangeHandler(e: any): void {
    const clonedOptions = questionsData.options;
    clonedOptions[e.target.id].correct = e.target.checked;
    setQuestionsData({
      ...questionsData,
      options: clonedOptions,
    });
  }

  function onGroupNameChangeHandler(e: any): void {
    setGroupName(e.target.value);
  }

  function onOptionChangeHandler(e: any): void {
    const clonedOptions = questionsData.options;
    clonedOptions[e.target.id].choice = e.target.value;
    setQuestionsData({
      ...questionsData,
      options: clonedOptions,
    });
  }

  const ComponentToShow = () => {
    const cookie = new Cookies();
    if (cookie.get("token") !== undefined) {
      return (
        <QuestionsView
          subjectNameChange={subjectNameChange}
          subjectName={subjectName}
          areaNameChange={areaNameChange}
          areaName={areaName}
          cancelUpdate={cancelUpdateEditingQuestion}
          updateQuestion={updateEditingQuestion}
          currentlyEditing={editingQuestionState}
          editQuestion={editQuestion}
          deleteQuestion={deleteQuestion}
          submitQuestions={submitQuestions}
          groupName={groupName}
          optionExplanation={onOptionExplanationHandlerChange}
          correctHandler={onCorrectChangeHandler}
          groupNameChange={onGroupNameChangeHandler}
          categoryNameChange={categoryNameChange}
          categoryName={category}
          addedQuestions={addedQuestions}
          optionChange={onOptionChangeHandler}
          changeName={onNameChangeHandler}
          saveQuestion={saveQuestion}
          newQuestion={questionsData}
          tagChange={onTagChangeHandler}
        />
      );
    } else {
      return <Redirect to={"/"} />;
    }
  };

  return <React.Fragment>{ComponentToShow()}</React.Fragment>;
};

export default QuestionsContainer;
