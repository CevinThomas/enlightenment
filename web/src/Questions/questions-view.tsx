import React, {MutableRefObject, useEffect} from 'react';
import styled from "styled-components";
import Heading from "../components/heading";
import QuestionCrud from "../components/questionCrud";
import {QuestionsData} from "../interfaces/questionsData";
import QuestionOverview from "../components/questionOverview";

const MainContainer = styled.main`

`

const HeadingSection = styled.section`

`

const FormsSection = styled.section`

`

const SubmitSection = styled.section`

`

const AddQuestionButton = styled.button`

`

const SubmitQuestions = styled.button`

`

const GroupNameInput = styled.input`

`

const SavedUpdatedQuestion = styled.button`

`

const CancelUpdatedQuestion = styled.button`

`

interface IProps {
    addedQuestions: Array<QuestionsData>
    newQuestion: QuestionsData,
    groupName: string,
    submitQuestions: () => void,
    saveQuestion: () => void,
    changeName: (e: any) => void,
    optionChange: (e: any) => void,
    categoryChange: (e: any) => void,
    groupNameChange: (e: any) => void,
    correctHandler: (e: any) => void,
    optionExplanation: (e: any) => void,
    deleteQuestion: (id: number) => void,
    editQuestion: (id: number) => void,
    currentlyEditing: boolean,
    updateQuestion: (id: number) => void,
    cancelUpdate: (id: number) => void
}

const QuestionsView = ({newQuestion, saveQuestion, changeName, optionChange, addedQuestions, categoryChange, groupNameChange, correctHandler, optionExplanation, groupName, submitQuestions, deleteQuestion, editQuestion, currentlyEditing, updateQuestion, cancelUpdate}: IProps) => {

    return (
        <MainContainer>
            <HeadingSection>
                <Heading>{currentlyEditing && newQuestion !== undefined ? `Currently Editing: ${newQuestion.name}`: "Add Questions"}</Heading>
            </HeadingSection>
            <FormsSection>
                {currentlyEditing && newQuestion !== undefined ? <React.Fragment><QuestionCrud optionExplanation={optionExplanation} correctHandler={correctHandler} categoryChange={categoryChange} optionChange={optionChange} changeName={changeName} questionData={newQuestion}/><SavedUpdatedQuestion onClick={() => updateQuestion(newQuestion.id)}>Update!</SavedUpdatedQuestion><CancelUpdatedQuestion onClick={() => cancelUpdate(newQuestion.id)}>Cancel</CancelUpdatedQuestion></React.Fragment> : <React.Fragment><GroupNameInput onChange={groupNameChange} value={groupName} placeholder={"Group Name"} />
                {addedQuestions.length !== 0 ? addedQuestions.map((question: QuestionsData) => {
                    return <QuestionOverview editQuestion={editQuestion} deleteQuestion={deleteQuestion} key={question.id} questionsData={question}/>
                }) : null}
                    <QuestionCrud optionExplanation={optionExplanation} correctHandler={correctHandler} categoryChange={categoryChange} optionChange={optionChange} changeName={changeName} questionData={newQuestion}/>
                    <AddQuestionButton onClick={saveQuestion}>Add Question</AddQuestionButton></React.Fragment> }

            </FormsSection>
            <SubmitSection>
                <SubmitQuestions onClick={submitQuestions}>Submit</SubmitQuestions>
            </SubmitSection>
        </MainContainer>
    );
};

export default QuestionsView;
