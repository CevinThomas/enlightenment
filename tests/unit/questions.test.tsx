import 'react-native';
import React from 'react';
import Questions from '../../src/views/questions';
import renderer from 'react-test-renderer';

test('Questions View Renders', () => {
    const tree = renderer.create(
        <Questions/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
