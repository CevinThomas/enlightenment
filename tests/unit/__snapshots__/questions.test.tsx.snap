// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions View Renders 1`] = `
<View>
  <View />
</View>
`;
